core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceUnitTests.scala

```diff
@@ -217,4 +224,298 @@ class WorkspaceServiceUnitTests extends AnyFlatSpec with OptionValues with Mocki
+      WorkspaceACLUpdate(readerEmail, WorkspaceAccessLevels.Write, Option(false), Option(false))
+    )
+
+    Await.result(service.updateACL(WorkspaceName("fake_namespace", "fake_name"), aclUpdates, true), Duration.Inf)
+
+    verify(samDAO, never).addUserToPolicy(any(), any(), any(), any(), any())
+    verify(samDAO, never).removeUserFromPolicy(any(), any(), any(), any(), any())
+    verify(wsmDAO).removeRole(ArgumentMatchers.eq(workspaceId),
+                              ArgumentMatchers.eq(WorkbenchEmail(writerEmail)),
+                              ArgumentMatchers.eq(IamRole.WRITER),
+                              any()
+    )
+    verify(wsmDAO).removeRole(ArgumentMatchers.eq(workspaceId),
+                              ArgumentMatchers.eq(WorkbenchEmail(readerEmail)),
+                              ArgumentMatchers.eq(IamRole.READER),
+                              any()
+    )
+    verify(wsmDAO).grantRole(ArgumentMatchers.eq(workspaceId),
+                             ArgumentMatchers.eq(WorkbenchEmail(readerEmail)),
+                             ArgumentMatchers.eq(IamRole.WRITER),
+                             any()
+    )
+  }
+
+  it should "not allow share writers for McWorkspaces" in {
+    val ownerEmail = "owner@example.com"
+    val writerEmail = "writer@example.com"
+    val readerEmail = "reader@example.com"
+    val workspaceId = UUID.randomUUID()
+
+    val wsmDAO = mockWsmForAclTests(ownerEmail, writerEmail, readerEmail)
+    val datasource = mockDatasourceForAclTests(WorkspaceType.McWorkspace, workspaceId)
+    val samDAO = mockSamForAclTests()
+
+    val aclUpdates = Set(
+      WorkspaceACLUpdate(writerEmail, WorkspaceAccessLevels.Write, Option(true), Option(false))
+    )
+
+    val service =
+      workspaceServiceConstructor(datasource, samDAO = samDAO, workspaceManagerDAO = wsmDAO)(defaultRequestContext)
+    val exception = intercept[RawlsExceptionWithErrorReport] {
+      Await.result(service.updateACL(WorkspaceName("fake_namespace", "fake_name"), aclUpdates, true), Duration.Inf)
+    }
+
+    exception.errorReport.statusCode shouldBe Option(StatusCodes.BadRequest)
+  }
+
+  it should "not allow share readers for McWorkspaces" in {
+    val ownerEmail = "owner@example.com"
+    val writerEmail = "writer@example.com"
+    val readerEmail = "reader@example.com"
+    val workspaceId = UUID.randomUUID()
+
+    val wsmDAO = mockWsmForAclTests(ownerEmail, writerEmail, readerEmail)
+    val datasource = mockDatasourceForAclTests(WorkspaceType.McWorkspace, workspaceId)
+    val samDAO = mockSamForAclTests()
+
+    val aclUpdates = Set(
+      WorkspaceACLUpdate(readerEmail, WorkspaceAccessLevels.Read, Option(true), Option(false))
+    )
+
+    val service =
+      workspaceServiceConstructor(datasource, samDAO = samDAO, workspaceManagerDAO = wsmDAO)(defaultRequestContext)
+    val exception = intercept[RawlsExceptionWithErrorReport] {
```

Comment:
Same comment about `RawlsExceptionWithErrorReport` -- I'd like to be more certain we're not covering up an unrelated error condition here.